zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -219,44 +267,152 @@ where
-            Some(idx) => idx,
-            None => panic!("Output 0 should exist in the transaction"),
-        },
-        RecipientAddress::Transparent(addr) => {
-            let script = addr.script();
-            tx.transparent_bundle()
-                .and_then(|b| {
-                    b.vout
-                        .iter()
-                        .enumerate()
-                        .find(|(_, tx_out)| tx_out.script_pubkey == script)
-                })
-                .map(|(index, _)| index)
-                .expect("we sent to this address")
+    let sent_outputs = request.payments.iter().enumerate().map(|(i, payment)| {
+        let idx = match &payment.recipient_address {
+            // Sapling outputs are shuffled, so we need to look up where the output ended up.
+            RecipientAddress::Shielded(_) =>
+                tx_metadata.output_index(i).expect("An output should exist in the transaction for each shielded payment."),
+            RecipientAddress::Transparent(addr) => {
+                let script = addr.script();
+                tx.transparent_bundle()
+                    .and_then(|b| {
+                        b.vout
+                            .iter()
+                            .enumerate()
+                            .find(|(_, tx_out)| tx_out.script_pubkey == script)
+                    })
+                    .map(|(index, _)| index)
+                    .expect("An output should exist in the transaction for each transparent payment.")
+            }
+        };
+
+        SentTransactionOutput {
+            output_index: idx,
+            recipient_address: &payment.recipient_address,
+            value: payment.amount,
+            memo: payment.memo.clone()
         }
-    };
+    }).collect();
+
+    wallet_db.store_sent_tx(&SentTransaction {
+        tx: &tx,
+        created: time::OffsetDateTime::now_utc(),
+        outputs: sent_outputs,
+        account,
+        utxos_spent: vec![],
+    })
+}
+
+#[cfg(feature = "transparent-inputs")]
+#[allow(clippy::too_many_arguments)]
+pub fn shield_funds<E, N, P, D, R>(
+    wallet_db: &mut D,
+    params: &P,
+    prover: impl TxProver,
+    account: AccountId,
+    sk: &secp256k1::SecretKey,
+    extsk: &ExtendedSpendingKey,
+    memo: &MemoBytes,
+    confirmations: u32,
+) -> Result<D::TxRef, E>
+where
+    E: From<Error<N>>,
+    P: consensus::Parameters,
+    R: Copy + Debug,
+    D: WalletWrite<Error = E, TxRef = R>,
+{
+    let (latest_scanned_height, latest_anchor) = wallet_db
+        .get_target_and_anchor_heights()
+        .and_then(|x| x.ok_or_else(|| Error::ScanRequired.into()))?;
+
+    // derive the corresponding t-address
+    let taddr = derive_transparent_address_from_secret_key(sk);
+
+    // derive own shielded address from the provided extended spending key
+    let z_address = extsk.default_address().unwrap().1;
+
+    let exfvk = ExtendedFullViewingKey::from(extsk);
+
+    let ovk = exfvk.fvk.ovk;
```

Comment:
Should this PR block on that being resolved?

Reply:
I can't remember whether we specified this (@str4d?), but since it's a matter of revealing more information than may have been intended, I think it's important to resolve.

Reply:
This will be resolved by #480